complement/pipeline.yml

```diff
@@ -37,3 +40,57 @@ steps:
       automatic:
         - exit_status: 128
           limit: 3
+
+  - command:
+      # Build the Synapse for Complement docker image, which is located at:
+      # https://github.com/matrix-org/complement/blob/master/dockerfiles/Synapse.Dockerfile.
+      - docker build -t complement-synapse -f dockerfiles/Synapse.Dockerfile dockerfiles/
+      # Run the tests!
+      - COMPLEMENT_BASE_IMAGE=complement-synapse go test -v -tags "synapse_blacklist" ./tests
+    label: "\U0001F9EA Complement / Synapse Monolith / :go: 1.15"
+    agents:
+      queue: "medium"
+    plugins:
+      - docker#v3.7.0:
+          # The dockerfile for this image is at https://github.com/matrix-org/complement/blob/master/dockerfiles/ComplementCIBuildkite.Dockerfile.
+          image: "matrixdotorg/complement:latest"
+          mount-buildkite-agent: false
+          # Complement needs to know if it is running under CI
```

Comment:
I agree that this could be used to diverge CI test runs from what's intended during development if this environment variable is abused. And we'll need to ensure this doesn't happen.

However at the moment the purpose of this variable is very simple. Some changes are needed if Complement itself is running in a docker container (and given access to the host's docker) versus running on the host directly. See where it is used:

Talking to the right docker runtime:

https://github.com/matrix-org/complement/blob/ee3715d9a2788263ac00990c260b2cf8a4f168ed/internal/docker/builder.go#L51-L56

Placement of CA files:

https://github.com/matrix-org/complement/blob/ee3715d9a2788263ac00990c260b2cf8a4f168ed/internal/docker/builder.go#L343-L408

https://github.com/matrix-org/complement/blob/ee3715d9a2788263ac00990c260b2cf8a4f168ed/internal/federation/server.go#L228-L245

This could possibly be implemented in a different manner than an environment variable, but I think the need for split code paths would remain.

Reply:
thanks, that's helpful.

Looking at this harder, I think the problem is really just that it's a crappy name for the env var. It's not "are we running under CI", but rather "is complement running under docker?" Presumably that's something you could detect (cf https://stackoverflow.com/questions/20010199/how-to-determine-if-a-process-runs-inside-lxc-docker).

To be clear: I'm not insisting you change it - I'm just saying it might be a thing to improve in the future.

Reply:
Good points. I've created an issue for this in Complement's repo: https://github.com/matrix-org/complement/issues/60

Suggest we do so after merging this.